### PR TITLE
feat!: Permissionless ICS 

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: #2108

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: x/ccv/provider/keeper/msg_server.go [hunk 1]
```diff
@@ -309,16 +307,163 @@ func (k msgServer) SetConsumerCommissionRate(goCtx context.Context, msg *types.M
...
+	// initialization parameters are optional and hence could be nil
+	if msg.InitializationParameters != nil {
+		if err := k.Keeper.SetConsumerInitializationParameters(ctx, consumerId, *msg.InitializationParameters); err != nil {
+			return &types.MsgCreateConsumerResponse{}, errorsmod.Wrapf(types.ErrInvalidConsumerInitializationParameters,
+				"cannot set consumer initialization parameters: %s", err.Error())
+		}
+	}
+
+	// power-shaping parameters are optional and hence could be null
+	if msg.PowerShapingParameters != nil {
```
Comment (mpoke) [inline x/ccv/provider/keeper/msg_server.go:334]: @insumity Update allowlist and denylist inside this block. See `UpdateConsumer`, i.e., 
```golang
    k.Keeper.UpdateAllowlist(ctx, consumerId, msg.PowerShapingParameters.Allowlist)
    k.Keeper.UpdateDenylist(ctx, consumerId, msg.PowerShapingParameters.Denylist)
```
It is fine to not update the min power for TopN (see `UpdateMinimumPowerInTopN`) as the chain is still in Registered phase. The min power will be computed and stored in state on `MakeConsumerGenesis` at spawn time. Maybe add this as a comment.

File: x/ccv/provider/proposal_handler.go
Comment (mpoke) [inline x/ccv/provider/proposal_handler.go:1]: @bermuell This entire file needs to be removed. This will affect the e2e tests though.

File: x/ccv/provider/keeper/msg_server.go [hunk 2]
```diff
@@ -293,14 +292,14 @@ func (k msgServer) SetConsumerCommissionRate(goCtx context.Context, msg *types.M
 		return nil, err
 	}
 
-	if err := k.HandleSetConsumerCommissionRate(ctx, msg.ChainId, types.NewProviderConsAddress(consAddr), msg.Rate); err != nil {
+	if err := k.HandleSetConsumerCommissionRate(ctx, msg.ConsumerId, types.NewProviderConsAddress(consAddr), msg.Rate); err != nil {
 		return nil, err
 	}
 
 	ctx.EventManager().EmitEvents(sdk.Events{
```
Comment (mpoke) [inline x/ccv/provider/keeper/msg_server.go:299]: @insumity Add more events for all new messages.

Comment (mpoke) [reply x/ccv/provider/keeper/msg_server.go:273]: Done in https://github.com/cosmos/interchain-security/pull/2221

**Review by insumity — APPROVED** [review]
As discussed with @mpoke. I did **not** review but accepting (✅) to unblock cutting. I'll review at a later stage.

**Review by MSalopek — APPROVED** [review]
Admin approval.
Review will follow.

File: proto/interchain_security/ccv/provider/v1/tx.proto [hunk 3]
```diff
@@ -23,34 +23,45 @@ service Msg {
...
   // The validator address on the provider
   string provider_addr = 2 [ (gogoproto.moretags) = "yaml:\"address\"" ];
   // The consensus public key to use on the consumer.
   // in json string format corresponding to proto-any, ex:
   // `{"@type":"/cosmos.crypto.ed25519.PubKey","key":"Ui5Gf1+mtWUdH8u3xlmzdKID+F3PK0sfXZ73GZ6q6is="}`
   string consumer_key = 3;
 
-  // Tx signer address
-  string signer = 4 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string submitter = 4 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```
Comment (MSalopek) [inline proto/interchain_security/ccv/provider/v1/tx.proto:61]: Revert the changes to not break compatibility. If you do this you will not be able to unmarshall historic Txs.

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+
+		// if current block time is equal to or after spawnTime, and the chain is initialized, we can launch the chain
+		consumerIds, err := k.GetConsumersToBeLaunched(ctx, spawnTime)
+		if err != nil {
+			k.Logger(ctx).Error("failed to retrieve consumers to launch",
+				"spawn time", spawnTime,
+				"error", err)
+			continue
+		}
+		if len(result)+len(consumerIds.Ids) >= int(limit) {
```
Comment (MSalopek) [inline x/ccv/provider/keeper/consumer_lifecycle.go:128]: Simplify this. This is causing a bug at the callsite.

At the callsite, you are removing ALL items under a spawn time.

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+		initializationParameters, err := k.GetConsumerInitializationParameters(ctx, consumerId)
+		if err != nil {
+			ctx.Logger().Error("could not retrieve initialization record",
+				"consumerId", consumerId,
+				"error", err)
+			continue
+		}
+		// Remove consumer to prevent re-trying launching this chain.
+		// We would only try to re-launch this chain if a new `MsgUpdateConsumer` message is sent.
+		err = k.RemoveConsumerToBeLaunched(ctx, consumerId, initializationParameters.SpawnTime)
```
Comment (MSalopek) [inline x/ccv/provider/keeper/consumer_lifecycle.go:76]: Here, you are removing ALL ids under a spawn time index. You should be removing a partial list.

See: [GetConsumersReadyToLaunch](https://github.com/cosmos/interchain-security/pull/2171/files/129164c31d796125fcab3d1d7ff54f941dc6b6a4#r1747167739)

File: x/ccv/provider/keeper/msg_server.go [hunk 6]
```diff
@@ -293,32 +313,342 @@ func (k msgServer) SetConsumerCommissionRate(goCtx context.Context, msg *types.M
...
+	}
+	if err := k.Keeper.SetConsumerInitializationParameters(ctx, consumerId, initializationParameters); err != nil {
+		return &resp, errorsmod.Wrapf(types.ErrInvalidConsumerInitializationParameters,
+			"cannot set consumer initialization parameters: %s", err.Error())
+	}
+	if !initializationParameters.SpawnTime.IsZero() {
+		// add SpawnTime event attribute
+		eventAttributes = append(eventAttributes,
+			sdk.NewAttribute(types.AttributeConsumerSpawnTime, initializationParameters.SpawnTime.String()))
+	}
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/msg_server.go:390]: Feels like this should go in line 409 after the If on InitializeConsumer and use the SpawnTime from there

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+			k.Logger(ctx).Error("failed to retrieve consumers to launch",
+				"spawn time", spawnTime,
+				"error", err)
+			continue
+		}
+		if len(result)+len(consumerIds.Ids) >= int(limit) {
+			remainingConsumerIds := len(result) + len(consumerIds.Ids) - int(limit)
+			if len(consumerIds.Ids[:len(consumerIds.Ids)-remainingConsumerIds]) == 0 {
+				return result
+			}
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/consumer_lifecycle.go:132]: This is complicated and uses so much index math, but it is just a check that len(result) == int(limit)

The math:
len(consumerIds.Ids) - remainingConsumerIds =
len(consumerIds.Ids) - len(result) - len(consumerIds.Ids) + int(limit) =
int(limit) - len(result)

the only way for consumerIds.Ids[:int(limit)-len(result)] to be empty is if int(limit) - len(result) == 0

Would suggest refactoring the surrounding if block with this in mind and to avoid the complicated index math

File: x/ccv/provider/types/msg.go [hunk 8]
```diff
@@ -3,521 +3,594 @@ package types
...
-	for _, denom := range msg.DenomsToAdd {
-		if !sdk.NewCoin(denom, math.NewInt(1)).IsValid() {
-			return fmt.Errorf("invalid change reward denoms proposal: %s is not a valid denom", denom)
-		}
+	return nil
+}
+
+func ValidateByteSlice(hash []byte, maxLength int) error {
+	if len(hash) > maxLength {
+		return fmt.Errorf("hash is too long; got: %d, max: %d", len(hash), MaxHashLength)
```
Comment (p-offtermatt) [inline x/ccv/provider/types/msg.go:569]: Tiny nit/bug here

```suggestion
		return fmt.Errorf("hash is too long; got: %d, max: %d", len(hash), maxLength)
```

File: proto/interchain_security/ccv/provider/v1/provider.proto [hunk 9]
```diff
@@ -362,3 +366,122 @@ message ConsumerRewardsAllocation {
...
 }
+
+// ConsumerMetadata contains general information about the registered chain
+message ConsumerMetadata {
+  // the name of the chain
+  string name = 1;
+  // the description of the chain
+  string description = 2;
+  // the metadata (e.g., GitHub repository URL) of the chain
+  string metadata = 3;
```
[the remark this answers is not in the document]
Comment (mpoke) [reply proto/interchain_security/ccv/provider/v1/provider.proto:377]: We leave it as a string for now.

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+		return err
+	}
+
+	// Set minimum height for equivocation evidence from this consumer chain
+	k.SetEquivocationEvidenceMinHeight(ctx, consumerId, initializationRecord.InitialHeight.RevisionHeight)
+
+	// Consumers start out with the unbonding period from the consumer addition prop
+	consumerUnbondingPeriod := initializationRecord.UnbondingPeriod
+
+	// Create client state by getting template client from parameters and filling in zeroed fields from proposal.
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/consumer_lifecycle.go:189]: Same, proposal seems outdated

File: x/ccv/provider/types/throttle.go [hunk 11]
```diff
@@ -1,17 +0,0 @@
-package types
-
-import (
-	"time"
-)
-
-// NewGlobalSlashEntry creates a new GlobalSlashEntry.
-func NewGlobalSlashEntry(recvTime time.Time, consumerChainID string,
```
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/provider/types/throttle.go:8]: No longer used after the introduction of throttling v2.

File: x/ccv/provider/keeper/consumer_lifecycle.go (hunk 5 above)
Comment (mpoke) [reply x/ccv/provider/keeper/consumer_lifecycle.go:76]: See https://github.com/cosmos/interchain-security/pull/2236

**Comment by p-offtermatt** [comment]
Also outdated here:
https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/params.go#L14-L18

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 12]
```diff
@@ -0,0 +1,670 @@
...
+	if err != nil {
+		return gen, nil, fmt.Errorf("unable to create validator set from updates computed from key assignment in MakeConsumerGenesis: %s", err)
+	}
+	hash := tmtypes.NewValidatorSet(updatesAsValSet).Hash()
+
+	consumerGenesisParams := ccv.NewParams(
+		true,
+		initializationRecord.BlocksPerDistributionTransmission,
+		initializationRecord.DistributionTransmissionChannel,
+		"", // providerFeePoolAddrStr,
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/consumer_lifecycle.go:334]: I feel like this needs a bit more of a comment here, looks a bit sketchy and like someone just forgot to uncomment this without any extra comment

File: x/ccv/provider/keeper/consumer_lifecycle.go (hunk 7 above)
Comment (mpoke) [reply x/ccv/provider/keeper/consumer_lifecycle.go:132]: This was fixed in https://github.com/cosmos/interchain-security/pull/2236

**Comment by mpoke** [comment]
@p-offtermatt I addressed comments until here in https://github.com/cosmos/interchain-security/pull/2237

File: x/ccv/provider/keeper/consumer_lifecycle.go [hunk 13]
```diff
@@ -0,0 +1,670 @@
...
+	k.DeleteDenylist(ctx, consumerId)
+	k.DeleteAllOptedIn(ctx, consumerId)
+	k.DeleteConsumerValSet(ctx, consumerId)
+
+	k.DeleteConsumerRewardsAllocation(ctx, consumerId)
+	k.DeleteConsumerRemovalTime(ctx, consumerId)
+
+	// TODO (PERMISSIONLESS) add newly-added state to be deleted
+	// Note that we do not delete ConsumerIdToChainIdKey and ConsumerIdToPhase, as well
+	// as consumer metadata, initialization and power-shaping parameters.
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/consumer_lifecycle.go:517]: Please add some minimal explanation of why we don't delete this here

Comment (mpoke) [reply x/ccv/provider/keeper/consumer_lifecycle.go:517]: see https://github.com/cosmos/interchain-security/pull/2237/commits/1ab477bc7061e89f51cac19fed9d287f12c630bf

**Comment by p-offtermatt** [comment]
There is a bug around the spawntime in UpdateConsumer; essentially it's possible fo a user to set a consumer chain that will attempt to launch in every block.

Assume we have a Registered consumer chain with a SpawnTime.
Now, a user sends a MsgUpdateConsumer with InitializationParameters where SpawnTime==0.
(This is not validated against in ValidateBasic https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/types/msg.go#L524)

The new InitializationParameters get set here: https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/msg_server.go#L507

However, the InitializeConsumer call here https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/msg_server.go#L567 will fail because the SpawnTime is zero, see this check https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/consumer_lifecycle.go#L54

Thus, the PrepareConsumerForLaunch call here https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/msg_server.go#L568 does not update the SpawnTime in the ConsumerToBeLaunched store.

So the InitializationParameters will say SpawnTime==0, but the store will have SpawnTime=oldSpawnTime.

During BeginBlock, when the consumers are launched, this RemoveConsumerToBeLaunched call here will thus not remove the consumer https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/consumer_lifecycle.go#L76 and error, which causes the loop to continue.

In the extreme, someone can create 200 such consumers, and those will be re-tried to be launched in every block and no other consumers will ever launch

**Comment by p-offtermatt** [comment]
Another find:
The owner can update launched consumer chains; in particular the owner can arbitrarily modify the InitializationParameters of the already launched chain. afaict this doesn't have any further impact, but it means the InitializationParameters returned by QueryConsumerChain can be arbitrary, and it would look like the chain launched with different parameters than it actually launched with (i.e. someone malicious could make it look like the chain had e.g. very large ConsumerRedistributionFraction

**Comment by mpoke** [comment]
> So the InitializationParameters will say SpawnTime==0, but the store will have SpawnTime=oldSpawnTime.
> 
> During BeginBlock, when the consumers are launched, this RemoveConsumerToBeLaunched call here will thus not remove the consumer
> 
> https://github.com/cosmos/interchain-security/blob/5b6252c7726809f0c56beb346de328a152f6d4fc/x/ccv/provider/keeper/consumer_lifecycle.go#L76
> 
> and error, which causes the loop to continue.
> In the extreme, someone can create 200 such consumers, and those will be re-tried to be launched in every block and no other consumers will ever launch

Nice catch @p-offtermatt. I do think that https://github.com/cosmos/interchain-security/pull/2236 somehow fixed this, but we should probably not have inconsistencies between the initialization params and the time queues used to store the chains to be launched.
